### PR TITLE
java: introduce proper types and minimise reliance on ByteBuffer/byte[]

### PR DESCRIPTION
**Pull request by axic**
Part of #548.

### CODE REVIEW

File: bindings/java/java/src/main/java/org/ethereum/evmc/Host.java [hunk 1]
```diff
@@ -31,7 +33,7 @@ static ByteBuffer get_storage(int context_index, byte[] address, byte[] key) {
         requireNonNull(
             getContext(context_index),
             "HostContext does not exist for context_index: " + context_index);
-    return context.getStorage(address, key);
+    return context.getStorage(new Addres(address), new Bytes32(key)).getByteBuffer();
```
Comment (atoulme) [inline bindings/java/java/src/main/java/org/ethereum/evmc/Host.java:36]: typo Addres

File: bindings/java/java/src/main/java/org/ethereum/evmc/HostContext.java [hunk 2]
```diff
@@ -44,7 +45,7 @@
    * @param value The value to be stored.
    * @return The effect on the storage item.
    */
-  int setStorage(byte[] address, byte[] key, byte[] value);
+  int setStorage(Address address, bytes32 key, Bytes32 value);
```
Comment (atoulme) [inline bindings/java/java/src/main/java/org/ethereum/evmc/HostContext.java:48]: bytes32 should be Bytes32

File: bindings/java/java/src/main/java/org/ethereum/evmc/types/Message.java [hunk 3]
```diff
@@ -0,0 +1,45 @@
+// EVMC: Ethereum Client-VM Connector API.
+// Copyright 2019-2020 The EVMC Authors.
+// Licensed under the Apache License, Version 2.0.
+package org.ethereum.evmc.types;
+
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+
+public class TxContext {
+  private final bool is32bit() {
```
Comment (atoulme) [inline bindings/java/java/src/main/java/org/ethereum/evmc/types/Message.java:10]: bool -> boolean
is32bit is not used.

**Comment by atoulme** [comment]
I really like this idea. Would remove all guessing. Feels like you have some coding issues.

Do you intend for implementors to create Address, Bytes32 Uint256 objects? If so, they'll need public constructors.

**Comment by axic** [comment]
@atoulme thanks! This is still work in progress, hence the draft mode, but I think this would be the safest way to handle things. Only issue is `Message` which is a mess due to the pointer thingie.

File: bindings/java/java/src/main/java/org/ethereum/evmc/types/Message.java [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+  public ByteBuffer getByteBuffer() {
+    // This is big endian by default, which is what we want.
+    return ByteBuffer.allocateDirect(expectedLength() + this.input.remaining())
+        .putInt(this.kind)
+        .putInt(this.flags)
+        .putInt(this.depth)
+        .putLong(this.gas)
+        .put(this.destination.getByteBuffer())
+        .put(this.sender.getByteBuffer())
+        // TODO: change this offset field based on endianness
```
Comment (axic) [inline bindings/java/java/src/main/java/org/ethereum/evmc/types/Message.java:40]: Actually all these helpers operate in **current endianess** (https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#putInt(int)) so all this should be fine.

File: bindings/java/java/src/main/java/org/ethereum/evmc/types/Message.java [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+  public Address destination;
+  public Address sender;
+  public ByteBuffer input;
+  public Uint256 value;
+  public Bytes32 create2_salt;
+
+  public Message() {}
+
+  public ByteBuffer getByteBuffer() {
+    // This is big endian by default, which is what we want.
```
Comment (axic) [inline bindings/java/java/src/main/java/org/ethereum/evmc/types/Message.java:32]: ByteBuffer is set to big-endian by default, but we want to set it to native endian (helper to set: https://docs.oracle.com/javase/7/docs/api/java/nio/ByteBuffer.html#order(java.nio.ByteOrder)).

We can use `ByteOrder` for that: https://docs.oracle.com/javase/6/docs/api/java/nio/ByteOrder.html#nativeOrder()

**Comment by axic** [comment]
Need to add Result class, and also pass Result/Message in HostContext.call.

**Comment by axic** [comment]
This currently contains the desired API from the Java perspective, yet to actually implement all of this. I'm leaning towards now that the C part would need to just receive an instance of Result/Message and manually read/write the fields and compose the C struct, as opposed to trying to construct the C struct from within Java.

**Comment by axic** [comment]
With #706 Java support was dropped -- closing this.

**Pull request closed without merging**